### DISCUSSION TITLE
multi level IL

### DISCUSSION THREAD

**Discussion by Mandofskii** [NationalSecurityAgency/ghidra#5573] [General]
**Is your feature request related to a problem? Please describe.**
I used Binary Ninja before I switch to Ghidra, Binary Ninja had a good feature called BNIL, It decompiles disassembled code to its IL(BNIL) and C pseudocode.
Why I say It was a good feature ? Cause you can see what is happening in code step by step through BNIL levels and it supports SSA that Ghidra doesnt.

**Describe the solution you'd like**
Implement a multi level iL in Ghidra or add multiple levels to existing P-Code IL.

**Describe alternatives you've considered**

**Additional context**

Recommended by @Alom4hdi in #5562 

![BNIL-overview](https://github.com/NationalSecurityAgency/ghidra/assets/59086060/309dd430-f496-4f4a-b2d4-fff6101db5b7)

![BNIL](https://github.com/NationalSecurityAgency/ghidra/assets/59086060/103a2e35-8a00-429a-8652-8bfac2c45800)

**Comment by bdemick** [comment]
Take a look at the documentation in the C++ decompiler part of the source tree (you’ll need to use doxygen to generate it). There are multiple refinement stages that PCode goes through that might provide some similar functionality. Using it is definitely not as friendly as Binary Ninja’s ILs.

**Comment by LukeSerne** [comment]
Implementing multiple IL levels would be a major change to the way the decompiler works, and as such, it seems unlikely to me that it will happen. Ghidra uses p-code ([link to documentation](https://htmlpreview.github.io/?https://raw.githubusercontent.com/NationalSecurityAgency/ghidra/master/GhidraDocs/languages/html/pcoderef.html)) as IL. After the p-code has been generated from the instructions using the information in the Sleigh specification file(s), it is processed by the decompiler. After the decompiler has done its analysis passes, the result is still p-code, but this "refined" form is often called "high p-code". This final version of p-code is then "translated" into the C-like language that is the output of the decompiler.

If you compile a debug version of the decompiler yourself, you can dump this high p-code, although there is an issue (#5545) to add this option to Ghidra's UI as well. Another way to gain insight into the state of the p-code at various stages during the decompilation process would be to use [the DecompVis tool](https://github.com/LukeSerne/DecompVis/) I created. It allows you to see how the data flow graph, recovered from the p-code, changes as the various optimisation rules inside the decompiler are applied.

Also, you mention:
> Cause you can see what is happening in code step by step through BNIL levels and it supports SSA that Ghidra doesnt.

I'd argue you can already see what is happening as the assembly instructions of the code under analysis execute. In the listing window, you can enable p-code to be shown. That shows the p-code operations generated for every assembly instruction (i.e. the non-optimised version). While this is not the most efficient way to represent what the code under analysis is doing, it certainly is helpful to understand it.

![image](https://github.com/NationalSecurityAgency/ghidra/assets/32435769/db8a6f46-6c20-46fe-9129-c6106781dd22)

The p-code is transformed into SSA form by the decompiler (specifically, the `heritage` rule). Unfortunately, this "low p-code in SSA form" is not available from outside the decompiler (see #5137). I think it'd be a good idea to create a feature request that requests this "lowest" p-code in SSA form to be exposed to the Java side. It's probably easier and less work to implement than multiple IL levels, while still being a big improvement and allowing new types of analysis (for example by hooking the raw pcode to an SMT solver to deobfuscate a binary, as suggested [in this blog post](https://www.msreverseengineering.com/blog/2019/4/17/an-abstract-interpretation-based-deobfuscation-plugin-for-ghidra)).

**Reply by nevesnunes** [reply]
@LukeSerne Phi-node placement, including renamed assignments, seem to get encoded by the decompiler and parsed on the Java side, but with a catch: Default optimizations apply constant propagation and other transformations that do not preserve these assignments (e.g. constant varnodes do not have a defining pcodeop). These optimizations are avoided on the Java side by setting the simplification style corresponding to action group `firstpass`, which mainly executes the `heritage` action.

For example, consider the following x86-16 program:
```
                   FUN_0000_0000
0000:0000 ba 01 00        MOV        DX,0x1
0000:0003 0b d2           OR         DX,DX
0000:0005 74 05           JZ         LAB_0000_000c
0000:0007 bb 01 00        MOV        BX,0x1
0000:000a eb 03           JMP        LAB_0000_000f
                   LAB_0000_000c  XREF[1]: 0000:0005(j)
0000:000c bb 02 00        MOV        BX,0x2
                   LAB_0000_000f  XREF[1]: 0000:000a(j)
0000:000f 89 d8           MOV        AX,BX
0000:0011 c3              RET
```

We expect that the basic block at 0000:000f has a phi-function for register BX due to assignments done at 0000:0007 and 0000:0000c. Phi-functions are encoded as `MULTIEQUAL` pcodeops.

On each pcodeop, the phi-node variables are defined on input/output varnodes' defining pcodeop sequence numbers (i.e. `varnode.getDef().getSeqnum()`).

We can [debug function decompilation](https://www.nccgroup.com/research/earlyremoval-in-the-conservatory-with-the-wrench-exploring-ghidra-s-decompiler-internals-to-make-automatic-p-code-analysis-scripts/#decompiler-internals) of action group `firstpass` using `decomp_dbg`:
```
restore /tmp/1.xml
load function FUN_0000_0000
trace address 0x0 0x11
option setaction firstpass
decompile
```

Output:
```
DEBUG 0: constbase
0x00000000:14: **
   0x00000000:14: DF(0x00000000:14) = #0x0:1

DEBUG 1: heritage
0x0000000f:15: **
   0x0000000f:15: BX(0x0000000f:15) = BX(0x00000007:b) ? BX(0x0000000c:13)
0x00000011:16: **
   0x00000011:16: IP(0x00000011:16) = SUB42(EIP(0x00000011:18),#0x0)
0x00000011:17: **
   0x00000011:17: EIP+2:2(0x00000011:17) = SUB42(EIP(i),#0x2)
0x00000011:18: **
   0x00000011:18: EIP(0x00000011:18) = CONCAT22(EIP+2:2(0x00000011:17),IP(0x00000011:f))
0x00000003:3: DX(0x00000003:3) = DX(free) | DX(free)
   0x00000003:3: DX(0x00000003:3) = DX(0x00000000:0) | DX(0x00000000:0)
0x00000003:4: SF(0x00000003:4) = DX(free) < #0x0:2
   0x00000003:4: SF(0x00000003:4) = DX(0x00000003:3) < #0x0:2
0x00000003:5: ZF(0x00000003:5) = DX(free) == #0x0:2
   0x00000003:5: ZF(0x00000003:5) = DX(0x00000003:3) == #0x0:2
0x00000003:6: u0x00049800:2(0x00000003:6) = DX(free) & #0xff:2
   0x00000003:6: u0x00049800:2(0x00000003:6) = DX(0x00000003:3) & #0xff:2
0x00000003:7: u0x00049900:1(0x00000003:7) = POPCOUNT(u0x00049800:2(free))
   0x00000003:7: u0x00049900:1(0x00000003:7) = POPCOUNT(u0x00049800:2(0x00000003:6))
0x00000003:8: u0x00049a00:1(0x00000003:8) = u0x00049900:1(free) & #0x1:1
   0x00000003:8: u0x00049a00:1(0x00000003:8) = u0x00049900:1(0x00000003:7) & #0x1:1
0x00000003:9: PF(0x00000003:9) = u0x00049a00:1(free) == #0x0:1
   0x00000003:9: PF(0x00000003:9) = u0x00049a00:1(0x00000003:8) == #0x0:1
0x00000005:a: goto Block_1:0x0000000c if (ZF(free) != 0) else Block_2:0x00000007
   0x00000005:a: goto Block_1:0x0000000c if (ZF(0x00000003:5) != 0) else Block_2:0x00000007
0x0000000f:d: AX(0x0000000f:d) = BX(free)
   0x0000000f:d: AX(0x0000000f:d) = BX(0x0000000f:15)
0x00000011:e: u0x00041c00(0x00000011:e) = segment(SS(free),SP(free))
   0x00000011:e: u0x00041c00(0x00000011:e) = segment(SS(i),SP(i))
0x00000011:f: IP(0x00000011:f) = *(ram,u0x00041c00(free))
   0x00000011:f: IP(0x00000011:f) = *(ram,u0x00041c00(0x00000011:e))
0x00000011:10: SP(0x00000011:10) = SP(free) + #0x2:2
   0x00000011:10: SP(0x00000011:10) = SP(i) + #0x2:2
0x00000011:11: EIP(0x00000011:11) = segment(CS(free),IP(0x00000011:16))
   0x00000011:11: EIP(0x00000011:11) = segment(CS(i),IP(0x00000011:16))
0x00000011:12: return(EIP(free))
   0x00000011:12: return(EIP(0x00000011:11))

DEBUG 2: segmentize
0x00000011:e: u0x00041c00(0x00000011:e) = segment(SS(i),SP(i))
   0x00000011:e: u0x00041c00(0x00000011:e) = segmentop(ram,SS(i),SP(i))
0x00000011:11: EIP(0x00000011:11) = segment(CS(i),IP(0x00000011:16))
   0x00000011:11: EIP(0x00000011:11) = segmentop(ram,CS(i),IP(0x00000011:16))
```

Despite `segmentize` running after `heritage`, it preserves SSA form.

Compare with the Java side results, given by script [DumpHighPCode.java](https://github.com/nevesnunes/env/blob/1449179fb2605d0db03b93ac1fb7013d4dcf1080/common/code/snippets/ghidra/DumpHighPCode.java):
```
=== decompile
0x0000000f:0: AX(0x0000000f:0) = COPY(0x1)
0x00000011:1: RETURN(0x0, AX(0x0000000f:0))
=== firstpass
0x00000000:0: DF(0x00000000:0) = COPY(0x0)
0x00000000:1: DX(0x00000000:1) = COPY(0x1)
0x00000003:2: CF(0x00000003:2) = COPY(0x0)
0x00000003:3: OF(0x00000003:3) = COPY(0x0)
0x00000003:4: DX(0x00000003:4) = INT_OR(DX(0x00000000:1), DX(0x00000000:1))
0x00000003:5: SF(0x00000003:5) = INT_SLESS(DX(0x00000003:4), 0x0)
0x00000003:6: ZF(0x00000003:6) = INT_EQUAL(DX(0x00000003:4), 0x0)
0x00000003:7: u0x00049800(0x00000003:7) = INT_AND(DX(0x00000003:4), 0xff)
0x00000003:8: u0x00049900(0x00000003:8) = POPCOUNT(u0x00049800(0x00000003:7))
0x00000003:9: u0x00049a00(0x00000003:9) = INT_AND(u0x00049900(0x00000003:8), 0x1)
0x00000003:10: PF(0x00000003:10) = INT_EQUAL(u0x00049a00(0x00000003:9), 0x0)
0x00000005:11: CBRANCH(0x0000000c, ZF(0x00000003:6))
0x00000007:0: BX(0x00000007:0) = COPY(0x1)
0x0000000a:1: BRANCH(0x0000000f)
0x0000000c:0: BX(0x0000000c:0) = COPY(0x2)
0x0000000f:0: BX(0x0000000f:0) = MULTIEQUAL(BX(0x00000007:0), BX(0x0000000c:0))
0x0000000f:1: AX(0x0000000f:1) = COPY(BX(0x0000000f:0))
0x00000011:2: u0x00041c00(0x00000011:2) = SEGMENTOP(0x2bada850, SS, SP)
0x00000011:3: r0x286:2(0x00000011:3) = SUBPIECE(EIP, 0x2)
0x00000011:4: IP(0x00000011:4) = LOAD(0x1a1, u0x00041c00(0x00000011:2))
0x00000011:5: EIP(0x00000011:5) = PIECE(r0x286:2(0x00000011:3), IP(0x00000011:4))
0x00000011:6: SP(0x00000011:6) = INT_ADD(SP, 0x2)
0x00000011:7: IP(0x00000011:7) = SUBPIECE(EIP(0x00000011:5), 0x0)
0x00000011:8: EIP(0x00000011:8) = SEGMENTOP(0x2bada850, CS, IP(0x00000011:7))
0x00000011:9: RETURN(EIP(0x00000011:8))
```

On `firstpass`, we can see the assignment of BX was renamed based on the expected phi-function:
```
0x0000000f:0: BX(0x0000000f:0) = MULTIEQUAL(BX(0x00000007:0), BX(0x0000000c:0))
0x0000000f:1: AX(0x0000000f:1) = COPY(BX(0x0000000f:0))
```

Besides SSA form being tied to function definitions, and optimized-out with the default simplification style, are there any other edge cases?
